skbio/stats/ordination/_principal_coordinate_analysis.py

```diff
@@ -62,37 +84,60 @@ def pcoa(distance_matrix):
-       to deal with that problem (see Legendre & Legendre 1998, \S
-       9.2.3), but none are currently implemented here.
-
-       However, a warning is raised whenever negative eigenvalues
-       appear, allowing the user to decide if they can be safely
-       ignored.
+    Principal Coordinate Analysis is sometimes known as
+    metric multidimensional scaling or classical scaling.
+
+    If the distance is not euclidean (for example if it is a
+    semimetric and the triangle inequality doesn't hold),
+    negative eigenvalues can appear. There are different ways
+    to deal with that problem (see Legendre & Legendre 1998, \S
+    9.2.3), but none are currently implemented here.
+
+    However, a warning is raised whenever negative eigenvalues
+    appear, allowing the user to decide if they can be safely
+    ignored.
     """
     distance_matrix = DistanceMatrix(distance_matrix)
 
-    E_matrix = e_matrix(distance_matrix.data)
-
-    # If the used distance was euclidean, pairwise distances
-    # needn't be computed from the data table Y because F_matrix =
-    # Y.dot(Y.T) (if Y has been centred).
-    F_matrix = f_matrix(E_matrix)
-
-    # If the eigendecomposition ever became a bottleneck, it could
-    # be replaced with an iterative version that computes the
-    # largest k eigenvectors.
-    eigvals, eigvecs = eigh(F_matrix)
-
-    # eigvals might not be ordered, so we order them (at least one
-    # is zero). cogent makes eigenvalues positive by taking the
+    # Center distance matrix, a requirement for PCoA here
+    matrix_data = center_distance_matrix(distance_matrix.data, inplace=inplace)
+
+    # If no dimension specified, by default will compute all eigenvectors
+    # and eigenvalues
+    if number_of_dimensions == 0:
+        if method == "fsvd" and matrix_data.shape[0] > 10:
+            warn("FSVD: since no value for number_of_dimensions is specified, "
+                 "PCoA for all dimensions will be computed, which may "
+                 "result in long computation time if the original "
+                 "distance matrix is large.", RuntimeWarning)
+
+        # distance_matrix is guaranteed to be square
+        number_of_dimensions = matrix_data.shape[0]
+    elif number_of_dimensions < 0:
+        raise ValueError('Invalid operation: cannot reduce distance matrix '
+                         'to negative dimensions using PCoA. Did you intend '
+                         'to specify the default value "0", which sets '
+                         'the number_of_dimensions equal to the '
+                         'dimensionality of the given distance matrix?')
+
+    # Perform eigendecomposition
+    if method == "eigh":
+        # eigh does not natively support specifying number_of_dimensions, i.e.
+        # there are no speed gains unlike in FSVD. Later, we slice off unwanted
+        # dimensions to conform the result of eigh
+        # to the specified number_of_dimensions.
```

Comment:
Can you remove the line break after eigh? There's still space in line 127 for the text in line 128.

Reply:
not quite enough space, but I've reformatted it.